### PR TITLE
New Way of Defining Middleware for Better Types

### PR DESCRIPTION
**Pull request by seveibar**
Check out `basic.test.ts`, basically we have a new way of defining middleware with a helper method that allows for dependencies to be specified and evaluated, and correctly types requests. 

```ts
  const mw1: Middleware<{ mw1_artifact: number }> = (next) => (req, res) => {
    req.mw1_artifact = 1
    next(req, res)
  }

  const mw2: Middleware<{ mw2_artifact: number }> = (next) => (req, res) => {
    req.mw2_artifact = 2
    next(req, res)
  }

  const mw3: Middleware<{ mw3_artifact: number }, { mw1_artifact: number }> =
    (next) => (req, res) => {
      req.mw1_artifact // this is defined because of the dependency def
      next(req, res)
    }

  wrappers(mw1, (req, res) => {
    req
  })

  wrappers(mw1, mw2, (req, res) => {
    req
  })

  wrappers(mw1, mw3, (req, res) => {
    req
  })

  // This should throw because mw3 depends on mw1
  // @ts-expect-error
  wrappers(mw2, mw3, (req, res) => {
    req
  })
```

### CODE REVIEW

**Comment by seveibar** [comment]
One thing I don't like this approach is it's not super composable, e.g. how would you allow someone to pass middlewares as an array to a function? I think we could do better if we figured out how to inductively define wrappers, or added some helpers types to address some common use cases (array of dependent middleware etc.)

File: wrappers.ts [hunk 1]
```diff
@@ -46,28 +47,281 @@ const withLoggedArguments =
...
+export const extendRequest = <T extends Req, K>(req: T, merge: K): T & K => {
+  for (const [key, v] of Object.entries(merge)) {
+    ;(req as any)[key] = v
+  }
+  return req as any
+}
+
+type Wrappers1 = <Mw1RequestContext, Mw1Dep>(
+  mw1: Middleware<Mw1RequestContext, Mw1Dep>,
+  endpoint: (req: Mw1RequestContext, res: any) => any
```
Comment (codetheweb) [inline wrappers.ts]: can we type `res` as `NextApiResponse` instead of `any`?

**Review by codetheweb — APPROVED** [review]
nice solution!

File: wrappers.ts (hunk 1 above)
Comment (seveibar) [reply wrappers.ts]: yep!

**Comment by codetheweb** [comment]
> how would you allow someone to pass middlewares as an array to a function?

I think you might be able to use `infer` but I'm not sure if that would be able to unwrap multiple generics

**Pull request merged**